### DISCUSSION TITLE
Why "uniqExactIf" is much slower than "uniqExact" ?

### DISCUSSION THREAD

**Discussion by avelychko12** [ClickHouse/ClickHouse#49865] [Q&A] [answered]
Results from ClickHouse Cloud(192gb, 16cpu) v23.3 but can be reproduced on local instance 
```sql
select uniqExact(number) from numbers_mt(1e8);
1 row in set. Elapsed: 0.791 sec. Processed 100.00 million rows, 800.00 MB (126.38 million rows/s., 1.01 GB/s.)

select uniqExactIf(number, 1) from numbers_mt(1e8);
1 row in set. Elapsed: 4.762 sec. Processed 100.00 million rows, 800.00 MB (21.00 million rows/s., 168.00 MB/s.)
```

```sql
select uniqExact(number) from numbers_mt(1e8) where number % 5 = 0;
1 row in set. Elapsed: 0.204 sec. Processed 100.00 million rows, 800.00 MB (490.28 million rows/s., 3.92 GB/s.)

select uniqExactIf(number, number % 5 = 0) from numbers_mt(1e8);
1 row in set. Elapsed: 1.157 sec. Processed 100.00 million rows, 800.00 MB (86.47 million rows/s., 691.73 MB/s.)
```
With `numbers` or `uniq/uniqIf` it is also slower, but not as much

**Comment by nickitat** [comment, marked as answer]
it slower because merging of aggregation states is not parallelised for `uniqExactIf` in contrast with `uniqExact`.

**Reply by avelychko12** [reply]
Hm, thanks. 
Is it applicable for all `-If` aggregations? 
Also, curious to know, is it a known "issue" that will be fixed in the future or kinda core concept that merging should not be parallelized for `-If`?

**Reply by nickitat** [reply]
`uniqExact` is parallelised in its own way, merging of states of other aggregate functions is not affected by this problem. I created an issue for this: https://github.com/ClickHouse/ClickHouse/issues/49885
